core/drivers/dra7_rng.c

```diff
@@ -0,0 +1,202 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <initcall.h>
+#include <io.h>
+#include <keep.h>
+#include <kernel/interrupt.h>
+#include <kernel/misc.h>
+#include <kernel/spinlock.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <rng_support.h>
+
+#define	RNG_OUTPUT_L            0x0000
+#define	RNG_OUTPUT_H            0x0004
+#define	RNG_STATUS              0x0008
+#  define RNG_READY             BIT(0)
+#  define SHUTDOWN_OFLO         BIT(1)
+#define	RNG_INTMASK             0x000C
+#define	RNG_INTACK              0x0010
+#define	RNG_CONTROL             0x0014
+#  define ENABLE_TRNG           BIT(10)
+#define	RNG_CONFIG              0x0018
+#define	RNG_ALARMCNT            0x001C
+#define	RNG_FROENABLE           0x0020
+#define	RNG_FRODETUNE           0x0024
+#define	RNG_ALARMMASK           0x0028
+#define	RNG_ALARMSTOP           0x002C
+#define	RNG_LFSR_L              0x0030
+#define	RNG_LFSR_M              0x0034
+#define	RNG_LFSR_H              0x0038
+#define	RNG_COUNT               0x003C
+#define	RNG_OPTIONS             0x0078
+#define	RNG_EIP_REV             0x007C
+#define	RNG_MMR_STATUS_EN       0x1FD8
+#define	RNG_REV                 0x1FE0
+#define	RNG_SYS_CONFIG_REG      0x1FE4
+#  define RNG_AUTOIDLE          BIT(0)
+#define	RNG_MMR_STATUS_SET      0x1FEC
+#define	RNG_SOFT_RESET_REG      0x1FF0
+#  define RNG_SOFT_RESET        BIT(0)
+#define	RNG_IRQ_EOI_REG         0x1FF4
+#define	RNG_IRQSTATUS           0x1FF8
+
+#define RNG_CONTROL_STARTUP_CYCLES_SHIFT        16
+#define RNG_CONTROL_STARTUP_CYCLES_MASK         GENMASK_32(31, 16)
+
+#define RNG_CONFIG_MAX_REFIL_CYCLES_SHIFT       16
+#define RNG_CONFIG_MAX_REFIL_CYCLES_MASK        GENMASK_32(31, 16)
+#define RNG_CONFIG_MIN_REFIL_CYCLES_SHIFT       0
+#define RNG_CONFIG_MIN_REFIL_CYCLES_MASK        GENMASK_32(7, 0)
+
+#define RNG_ALARMCNT_ALARM_TH_SHIFT             0
+#define RNG_ALARMCNT_ALARM_TH_MASK              GENMASK_32(7, 0)
+#define RNG_ALARMCNT_SHUTDOWN_TH_SHIFT          16
+#define RNG_ALARMCNT_SHUTDOWN_TH_MASK           GENMASK_32(20, 16)
+
+#define RNG_CONTROL_STARTUP_CYCLES              0xff
+#define RNG_CONFIG_MIN_REFIL_CYCLES             0x21
+#define RNG_CONFIG_MAX_REFIL_CYCLES             0x22
+#define RNG_ALARM_THRESHOLD                     0xff
+#define RNG_SHUTDOWN_THRESHOLD                  0x4
+
+#define RNG_FRO_MASK    GENMASK_32(23, 0)
+
+#define RNG_REG_SIZE    0x2000
+
+register_phys_mem(MEM_AREA_IO_SEC, RNG_BASE, RNG_REG_SIZE);
+
+static unsigned int rng_lock = SPINLOCK_UNLOCK;
+
+uint8_t hw_get_random_byte(void)
+{
+	static int pos;
+	static union {
+		uint32_t val[2];
+		uint8_t byte[8];
+	} random;
+	vaddr_t rng = (vaddr_t)phys_to_virt(RNG_BASE, MEM_AREA_IO_SEC);
+	uint8_t ret;
+
+	cpu_spin_lock(&rng_lock);
```

Comment:
How could I miss that since I wrote https://github.com/OP-TEE/optee_os/commit/01f9de219d25f1bb3f880c7ab4cf15a4e2fcb3ef?... :o
Which makes me wonder: should we enable `CFG_TEE_CORE_DEBUG` by default in `mk/config.mk`?